mantidproject/mantid
# Add simplified sequential fitting to EngDiff UI fitting tab.

**Description of work.**

Add simplified sequential fitting to EngDiff UI fitting tab. 
The primary log (specified in the settings) dictates the order in which to fit the workspaces (choosing a blank primary log will fit in the order in which they appear in the UI table). In the settings it is also possible to choose to to fit in ascending or descending order of the primary log value.

Additionally the settings has been refactored - clicking Cancel should not update the settings. When OK or Apply is clicked then the settings are saved to file and should still persist across sessions. 

**To test:**
In the Engineering diffraction UI
(1) Open the settings (cog symbol in bottom left) and choose a primary log (the combo box at bottom - should be brief description above it)
(2) Choose a log (or leave blank in which case the runs will be fitted in the order they appear in the table of the fitting tab) - the runs provided are d=taken at different stress and strain. Choose an order by checking one of the ascending/descending boxes (selecting one should automatically update the check state of the other, they should never both have the same state). Click OK. 
(3) Open the settings again - change any of them but this time press Cancel. When you re-open the settings the changes should not have been saved.
(4) Load the following focussed data [ENGINX_data.zip](https://github.com/mantidproject/mantid/files/5392081/ENGINX_data.zip) but do not check the add to plot
(5) Check the Sequential Fit button is disabled (is greyed out)
(6) Add one of the runs to the plot by checking the appropriate box in the table
(7) Open the fit browser by clicking Fit on the plot toolbar - the Sequential Fit button should still be disabled.
(8) Add a function (e.g. couple of peaks and background) - the Sequential Fit button should now be enabled (it is enabled when the Fit button in the Fit menu of the browser is enabled).
(9) Add constraints (fix/tie parameters), you can also try with/without excluded regions (note the regions are not displayed in the plot - and will not be visible if you plot guess either)
(10) Make a note of the starting parameters and click the Sequential Fit button - check the following
- For each matrix workspace in the group with suffix "_fit" (one per fit parameter) there are no rows with NaNs (this indicates the same function was fitted to all). Check this also in the model table workspace in the same grouping.
- The parameters displayed in the fit browser should match those of the last fitted run (which can be inferred from the primary log values in the table workspace (grouped together with suffix _logs).
- Inspect the saved setups in the fit browser (Setup > Custom Setup) - there should be a model with the name of each workspace. Select each model and check the parameters match those in the tables in the ADS.

(11) Run the sequential fit again (you can change the model but you don't have to) - you should not get prompted to ask whether you want to overwrite the saved setup (this is a change of behaviour introduced in this PR)
(12) Repeat steps (2-11) with different logs (incl. blank) the order can be inferred from the last fitted run (the actual sorting is covered by a unit test)


Fixes #29828 

Don't merge until PR #29667 is merged





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
